File: ios/Yomiyasu/Core/Services/ProgressAPI.swift
import Foundation

struct ReadProgressRequest: Encodable, Sendable {
    let book: String
    let time: Int?
    let currentPage: Int
    let characters: Int
    let status: String
    var endDate: String? = nil
}

@MainActor
struct ProgressAPI {
    let client: APIClient

    func progress(forBook bookId: String) async throws -> ReadProgress? {
        let progress: ReadProgress = try await client.send(
            .get("api/readprogress", queryItems: [URLQueryItem(name: "book", value: bookId)])
        )
        return progress.id == nil ? nil : progress
    }

    func save(_ request: ReadProgressRequest) async throws {
        try await client.send(try Endpoint.post("api/readprogress", json: request))
    }

    func markCompleted(_ book: Book, currentPage: Int? = nil, characters: Int? = nil) async throws {
        let request = ReadProgressRequest(
            book: book.id,
            time: nil,
            currentPage: currentPage ?? book.pages ?? 0,
            characters: characters ?? book.characters ?? 0,
            status: "completed",
            endDate: ISO8601DateFormatter().string(from: .now)
        )
        try await save(request)
    }

    func markUnread(_ book: Book) async throws {
        let request = ReadProgressRequest(
            book: book.id,
            time: nil,
            currentPage: 0,
            characters: 0,
            status: "unread"
        )
        try await save(request)
    }

    func neighboringBook(of bookId: String, forward: Bool) async throws -> Book? {
        let action = forward ? "next" : "prev"
        let data = try await client.sendData(.get("api/books/\(bookId)/\(action)"))

        struct Probe: Decodable { let _id: String }

        let decoder = JSONDecoder.yomiyasu()

        if let probe = try? decoder.decode(Probe.self, from: data),
           probe._id == "end" || probe._id == "start" {
            return nil
        }

        return try decoder.decode(Book.self, from: data)
    }

    func stats() async throws -> UserStats {
        try await client.send(.get("api/readprogress/mystats"))
    }

    func graphs() async throws -> MonthlyGraphs {
        try await client.send(.get("api/readprogress/mygraphs"))
    }

    func streak(year: Int, month: Int) async throws -> [StreakDay] {
        // El endpoint usa meses base 0 (`new Date(year, month)` en el backend),
        // a diferencia de `logs/:year/:month/:day`, que usa base 1 ($month de Mongo).
        try await client.send(.get("api/readprogress/streak/\(year)/\(month - 1)"))
    }

    func logs(year: Int, month: Int, day: Int) async throws -> [ProgressRecord] {
        try await client.send(.get("api/readprogress/logs/\(year)/\(month)/\(day)"))
    }

    func all(page: Int, limit: Int, sort: String) async throws -> (records: [ProgressRecord], total: Int) {
        struct Page: Decodable {
            let data: [ProgressRecord]
            let total: Int
        }

        do {
            let response: Page = try await client.send(
                .get(
                    "api/readprogress/all",
                    queryItems: [
                        URLQueryItem(name: "page", value: String(page)),
                        URLQueryItem(name: "limit", value: String(limit)),
                        URLQueryItem(name: "sort", value: sort),
                    ]
                )
            )
            return (response.data, response.total)
        } catch let error as APIError {
            if case .http(let status, _) = error, status == 400 {
                return ([], 0)
            }
            throw error
        }
    }

    func speed(serieId: String) async throws -> [ProgressRecord] {
        try await client.send(.get("api/readprogress/serie/\(serieId)/speed"))
    }
}

File: android/app/src/main/java/es/manabe/yomiyasu/core/services/ProgressApi.kt
package es.manabe.yomiyasu.core.services

import es.manabe.yomiyasu.core.models.Book
import es.manabe.yomiyasu.core.models.MonthlyGraphs
import es.manabe.yomiyasu.core.models.ProgressRecord
import es.manabe.yomiyasu.core.models.ReadProgress
import es.manabe.yomiyasu.core.models.StreakDay
import es.manabe.yomiyasu.core.models.UserStats
import es.manabe.yomiyasu.core.networking.ApiClient
import es.manabe.yomiyasu.core.networking.ApiException
import es.manabe.yomiyasu.core.networking.Endpoint
import es.manabe.yomiyasu.core.networking.jsonBody
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.builtins.ListSerializer
import java.time.Instant
import javax.inject.Inject
import javax.inject.Singleton

@Serializable
data class ReadProgressRequest(
    val book: String,
    val time: Int? = null,
    val currentPage: Int,
    val characters: Int,
    val status: String,
    val endDate: String? = null,
)

@Serializable
private data class ProgressPage(
    val data: List<ProgressRecord> = emptyList(),
    val total: Int = 0,
)

@Serializable
private data class Probe(@SerialName("_id") val id: String)

@Singleton
class ProgressApi @Inject constructor(
    private val api: ApiClient,
) {

    suspend fun progressForBook(bookId: String): ReadProgress? {
        val progress = api.send(
            Endpoint.get("api/readprogress", listOf("book" to bookId)),
            ReadProgress.serializer(),
        )
        return if (progress.id == null) null else progress
    }

    suspend fun save(request: ReadProgressRequest) {
        api.send(
            Endpoint.post("api/readprogress", body = jsonBody(request)),
            es.manabe.yomiyasu.core.models.EmptyResponse.serializer(),
        )
    }

    suspend fun markCompleted(book: Book, currentPage: Int? = null, characters: Int? = null) {
        save(
            ReadProgressRequest(
                book = book.id,
                time = null,
                currentPage = currentPage ?: book.pages ?: 0,
                characters = characters ?: book.characters ?: 0,
                status = "completed",
                endDate = Instant.now().toString(),
            ),
        )
    }

    suspend fun markUnread(book: Book) {
        save(
            ReadProgressRequest(
                book = book.id,
                time = null,
                currentPage = 0,
                characters = 0,
                status = "unread",
            ),
        )
    }

    suspend fun neighboringBook(bookId: String, forward: Boolean): Book? {
        val action = if (forward) "next" else "prev"
        val data = api.sendBytes(Endpoint.get("api/books/$bookId/$action"))

        try {
            val probe = api.json.decodeFromString(Probe.serializer(), data.decodeToString())
            if (probe.id == "end" || probe.id == "start") return null
        } catch (_: Exception) {
            // No era un centinela: se intenta decodificar como libro
        }

        return try {
            api.json.decodeFromString(Book.serializer(), data.decodeToString())
        } catch (error: Exception) {
            throw ApiException.Decoding(error)
        }
    }

    suspend fun stats(): UserStats =
        api.send(Endpoint.get("api/readprogress/mystats"), UserStats.serializer())

    suspend fun graphs(): MonthlyGraphs =
        api.send(Endpoint.get("api/readprogress/mygraphs"), MonthlyGraphs.serializer())

    /**
     * El endpoint usa meses base 0 (`new Date(year, month)` en el backend),
     * a diferencia de `logs/:year/:month/:day`, que usa base 1 ($month de Mongo).
     */
    suspend fun streak(year: Int, month: Int): List<StreakDay> = api.send(
        Endpoint.get("api/readprogress/streak/$year/${month - 1}"),
        ListSerializer(StreakDay.serializer()),
    )

    suspend fun logs(year: Int, month: Int, day: Int): List<ProgressRecord> = api.send(
        Endpoint.get("api/readprogress/logs/$year/$month/$day"),
        ListSerializer(ProgressRecord.serializer()),
    )

    suspend fun all(page: Int, limit: Int, sort: String): Pair<List<ProgressRecord>, Int> {
        return try {
            val response = api.send(
                Endpoint.get(
                    "api/readprogress/all",
                    listOf(
                        "page" to page.toString(),
                        "limit" to limit.toString(),
                        "sort" to sort,
                    ),
                ),
                ProgressPage.serializer(),
            )
            response.data to response.total
        } catch (error: ApiException) {
            if (error is ApiException.Http && error.status == 400) {
                emptyList<ProgressRecord>() to 0
            } else {
                throw error
            }
        }
    }

    suspend fun speed(serieId: String): List<ProgressRecord> = api.send(
        Endpoint.get("api/readprogress/serie/$serieId/speed"),
        ListSerializer(ProgressRecord.serializer()),
    )
}
